pkg/scheduler/flavorassigner/flavorassigner.go

```diff
@@ -466,23 +477,69 @@ func (a *FlavorAssigner) assignFlavors(log logr.Logger, counts []int32) Assignme
 			}
 		}
 
-		for resName := range podSet.Requests {
-			if _, found := psAssignment.Flavors[resName]; found {
+		groupKey := strconv.Itoa(i)
+		tr := a.wl.Obj.Spec.PodSets[i].TopologyRequest
+		if tr != nil && tr.PodSetGroup != nil {
+			groupKey = *tr.PodSetGroup
+		}
+
+		if !slices.Contains(groupsOrder, groupKey) {
+			groupsOrder = append(groupsOrder, groupKey)
+		}
+
+		groupedRequests[groupKey] = append(groupedRequests[groupKey], IndexedPodSet{OriginalIndex: i, PodSet: &podSet, PodSetAssignment: &psAssignment})
+	}
+
+	for _, groupKey := range groupsOrder {
+		podSets := groupedRequests[groupKey]
+		requests := make(resources.Requests)
+		psIDs := make([]int, len(podSets))
+		for idx, podset := range podSets {
+			psIDs[idx] = podset.OriginalIndex
+			requests.Add(podset.PodSet.Requests)
+		}
+
+		groupFlavors := make(ResourceAssignment)
+		for _, ips := range podSets {
+			for resName := range ips.PodSetAssignment.Flavors {
+				groupFlavors[resName] = ips.PodSetAssignment.Flavors[resName]
+				break
+			}
+		}
+		var groupStatus *Status
+
+		for resName := range requests {
+			if _, found := groupFlavors[resName]; found {
 				// This resource got assigned the same flavor as its resource group.
 				// No need to compute again.
 				continue
 			}
-			flavors, status := a.findFlavorForPodSetResource(log, i, podSet.Requests, resName, assignment.Usage.Quota)
+			flavors, status := a.findFlavorForPodSetResource(log, psIDs, requests, resName, assignment.Usage.Quota)
 			if status.IsError() || len(flavors) == 0 {
-				psAssignment.Flavors = nil
-				psAssignment.Status = status
+				groupFlavors = nil
+				groupStatus = status
 				break
 			}
-			psAssignment.append(flavors, status)
+			maps.Copy(groupFlavors, flavors)
+			if groupStatus == nil {
+				groupStatus = status
+			} else if status != nil {
+				groupStatus.reasons = append(groupStatus.reasons, status.reasons...)
+			}
```

Comment:
Why is this `append` removed? It seems it nicely encapsulated the technical detail of lazy initialization. Maybe the name was suboptimal?

Reply:
I removed it, because I am no longer using `psAssignment` to keep flavors and statuses. I could create a separate struct such as `groupAssignment` containing `groupFlavors` and `groupStatus` and implement similar `append` method on that. WDYT?

Reply:
Let's keep as is in this PR.

However,  I think it will be much easier going forward if Status was always initialized in psAssignment. There are many places in code which need to now handle the fallback when Status is nil, this is error prone. I will open an issue for it. wdyt?

Reply:
Sure, let's do it that way. I think we could also introduce something like PodSetGroupAssignment, which would include all PodSetAssignments and update them as we update the Grouped one. That way the code here would be clean and at the same time we would have all Statuses in all podsets up-to-date all the time.